docs/blogs/2023/2023-10-28-numba-fp16.md

```diff
@@ -0,0 +1,116 @@
+
+\\[\textnormal{Joint tensor size} = \: B \times T \times U \times V \times 2 \times 4 \,\, \textnormal{bytes}\\] 
+
+Here, $B$ is the batch size, $T$ is the audio sequence length, $U$ is the text sequence length and $V$ is the vocabulary size. We multiply by 2 so we get the size of both the activations and the gradients. We then multiply by 4 because we assume an FP32 datatype (and a single FP32 value occupies 4 bytes).
+
+The audio waveform signal is commonly converted to 100 Hz spectrogram frames, which means each second of audio corresponds to 100 audio frames. Thus, for a single 20-second audio clip with about 100 subwords in its transcript, and a vocabulary of 1024 subword tokens, the size of the tensor would be ~1.6 Gigabytes:
+
+\\[\textnormal{Joint tensor size} = \: B \times \phantom{....}T\phantom{....}\times \phantom{.}U\phantom{.} \times \phantom{..}V\phantom{.} \times 2 \times 4 \,\, \textnormal{bytes}\phantom{=1.6 \textnormal{ Gigabytes}}\\] 
+\\[\phantom{\textnormal{Joint tensor size}} = \: 1 \times (20 \times 100)  \times 100 \times 1024 \times 2 \times 4 \,\, \textnormal{bytes}=1.6 \textnormal{ Gigabytes}\\] 
+
+This number is for a single audio sample. If we use a larger batch size, e.g. 10, for training, we will quickly run out of memory even on 16 GB GPUs. Also, remember, this is just the size of the joint network tensor: there is additional memory required to keep the model in memory, and to calculate the activation and gradients of the rest of the network!
+
+## <i><b>Enter Numba support for FP16 datatype</b></i>
+
+As of [Numba 0.57](https://numba.readthedocs.io/en/stable/release-notes.html#version-0-57-0-1-may-2023) release, FP16 datatype format is now supported natively. Using this, we can effectively halve the memory requirement of the above joint network tensor and support larger batch sizes with almost no changes to our NeMo workflow!
+
+NeMo utilizes [Numba's](https://numba.readthedocs.io/en/stable/index.html) [Just-in-time compile CUDA kernels](https://numba.pydata.org/numba-doc/latest/cuda/kernels.html) written in Python in order to efficiently compute the RNN-T loss (which requires manipulation of the joint network tensor). This allows a user to simply have Numba installed on their system, and without explicit compilation of C++ / CUDA code, they can train their RNN-T models easily. Furthermore, since the kernels are written in Python, it allows for simple modifications by researchers to develop advanced features such as [FastEmit](https://arxiv.org/abs/2010.11148), and even other extensions to the Transducer loss, such as [Token-and-Duration Transducers](https://arxiv.org/abs/2304.06795).
```

Comment:
This is the old documentation for Numba - current docs are at https://numba.readthedocs.io/en/latest/cuda/kernels.html

Reply:
Done